bayes_kit/rhat.py

```diff
@@ -29,10 +29,10 @@ def rhat(chains: list[SeqType]) -> FloatType:
     """
     if len(chains) < 2:
         raise ValueError(f"rhat requires len(chains) >= 2, but {len(chains) = }")
-    chain_lengths = [len(chain) for chain in chains]
+    chain_lengths = [len(np.asarray(chain)) for chain in chains]
     mean_chain_length = np.mean(chain_lengths)
-    means = [np.mean(chain) for chain in chains]
-    vars = [np.var(chain, ddof=1) for chain in chains]
+    means = [np.mean(np.asarray(chain)) for chain in chains]
+    vars = [np.var(np.asarray(chain), ddof=1) for chain in chains]
```

Comment:
To avoid doing the (possibly non-zero-cost) cast 3 times per chain, we should do this with a standard for loop rather than a series of comprehensions

Reply:
Is that going to be faster?  I thought for loops were to be avoided at all costs in Python.  Should I just create the right type

```
[np.asarray(chain) for chain in chains]
```

and then just apply to that?  And maybe I should be doing `chain.mean()` instead of `np.mean(chain)`?  Maybe it doesn't matter here, but everywhere I've looked, I've seen the former recommended.

Reply:
> I thought for loops were to be avoided at all costs in Python. 

I’ve never heard this advice given out generically. When dealing with numpy specifically, if there is a way which avoids a loop it will be faster, but this doesn’t count - the comprehension is still, essentially, just a loop and doesn’t have any preferable speed characteristics that I am aware of


In general yes, if there is a method on the type, it’s preferable as opposed to the np.name style. Part of the reason for this is that then Jax arrays and similar “just work”

Reply:
I just comprehended the chains instead.  If that's not good enough, we can profile.